kube-runtime/src/utils/event_modify.rs

```diff
@@ -0,0 +1,72 @@
+use core::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::task::ready;
+
+use futures::{Stream, TryStream};
+use pin_project::pin_project;
+
+use crate::watcher::{Error, Event};
+
+#[pin_project]
+/// Stream returned by the [`modify`](super::WatchStreamExt::modify) method.
+/// Modifies the [`Event`] item returned by the inner stream by calling
+/// [`modify`](Event::modify()) on it.
+pub struct EventModify<St, F> {
+    #[pin]
+    stream: St,
+    f: F,
+}
+
+impl<St, F, K> EventModify<St, F>
+where
+    St: TryStream<Ok = Event<K>>,
+    F: FnMut(&mut K),
+{
+    pub(super) fn new(stream: St, f: F) -> EventModify<St, F> {
+        Self { stream, f }
+    }
+}
+
+impl<St, F, K> Stream for EventModify<St, F>
+where
+    St: Stream<Item = Result<Event<K>, Error>>,
+    F: FnMut(&mut K),
+{
+    type Item = Result<Event<K>, Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut me = self.project();
+        Poll::Ready(match ready!(me.stream.as_mut().poll_next(cx)) {
+            Some(Ok(event)) => Some(Ok(event.modify(me.f))),
+            Some(Err(err)) => Some(Err(err)),
+            None => return Poll::Ready(None),
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod test {
+    use std::task::Poll;
+
+    use super::{Error, Event, EventModify};
+    use futures::{pin_mut, poll, stream, StreamExt};
+
+    #[tokio::test]
+    async fn eventmodify_modifies_innner_value_of_event() {
+        let st = stream::iter([Ok(Event::Applied(0)), Err(Error::TooManyObjects)]);
+        let ev_modify = EventModify::new(st, |x| {
+            *x += 1;
+        });
```

Comment:
good test!

can i request one more test on `[Ok(Event::Restarted(...))]` where we also check the stream end case (`assert!(matches!(poll!(ev_stream.next()), Poll::Ready(None)))`) as it shows up as uncovered atm.